direct/core/src/main/java/cz/o2/proxima/direct/commitlog/RetryableLogObserver.java

```diff
@@ -64,14 +65,23 @@ public final boolean onNext(StreamElement ingest, OnNextContext context) {
 
   @Override
   public boolean onError(Throwable throwable) {
-    if (delegate.onError(throwable)) {
-      numFailures++;
+    if (delegate.onError(throwable)
+        && (retryErrors || !(throwable instanceof Error))
+        && ++numFailures <= maxRetries) {
+
       log.error(
           "Error in observer {}, retry {} out of {}", name, numFailures, maxRetries, throwable);
-      return numFailures <= maxRetries;
+      return true;
     }
     log.error("Error in observer {} (non-retryable)", name, throwable);
-    return false;
+    onRetriesExhausted.accept(throwable);
```

Comment:
```
  * @param onRetriesExhausted handler for {@link Throwable} caught during processing and not
   *     retried
```

This seems weird, `handler` should be allowed to handle the exception. Shouldn't this code path look like this instead?
```
    onRetriesExhausted.accept(throwable);
    return false;
```

And default onRetriesExhausted impl would rethrow errors:
```
    if (throwable instanceof Error) {
      throw (Error) throwable;
    }
    if (throwable instanceof RuntimeException) {
      throw (RuntimeException) throwable;
    }
    throw new IllegalStateException("Retries exhausted retrying observer " + name, throwable);
    ```

Current impl. disallows user to just stop processing and handle the error in some other way (eg. failing readiness probe), without propagating the error.

Reply:
This is a good question. I'm not sure what is better approach here. Generally yes, we could allow users to gracefully handle the failure (not sure how the user could handle that, because adding additional retry seems useless - that could be done by increasing the number of retries - so perhaps some sort of circuit-breaker would be an option), but - at the same time - we must be sure, that we don't simply stop processing without explicit permission to do so.

So, thinking about that - I think we could make the `Consumer` a `UnaryFunction<Throwable, Resolution>`, where `Resultion` would be an `enum`
```java
enum Resolution {
  STOP_PROCESSING,
  RETHROW_EXCEPTION,
  IMMEDIATELY_EXIT
}
```
something like this. Returning `boolean` is not explicit enough, imo.

Reply:
👍 I like the approach.

Maybe `IMMEDIATELY_EXIT` can be renamed to `SHUTDOWN_JVM` / `EXIT_IMMEDIATELY`

Just a minor thing, let's exchange `UnaryFunction` for `Function`, there is no reason this should be serializable (same applies to the whole LogObserver interface)

Reply:
I tend to use functional interfaces consistently. I don't like mixing UnaryFunction and Function within the same codebase. Even if there is no actual need for i ti to be Serializable (actually, LogObserver currently *is* Serializable - though it need not be - and therefore having non-Serializable Function would cause warnings from static code analyzer).